File: app/src/main/java/com/mesada/shop/repository/CartRepository.java
package com.mesada.shop.repository;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.mesada.shop.model.Cart;
import com.mesada.shop.model.Product;

import java.util.ArrayList;
import java.util.List;

public class CartRepository {

    private MutableLiveData<List<Cart>> mutableCart = new MutableLiveData<>();
    private MutableLiveData<Double> mutableTotalPrice = new MutableLiveData<>();

    public LiveData<List<Cart>> getCart() {
        if (mutableCart.getValue() == null) {
            initCart();
        }
        return mutableCart;
    }

    public void initCart() {
        mutableCart.setValue(new ArrayList<Cart>());
        calculateCartTotal();
    }

    public boolean addItemToCart(Product product) {
        if (mutableCart.getValue() == null) {
            initCart();
        }
        List<Cart> cartItemList = new ArrayList<>(mutableCart.getValue());
        for (Cart cartItem: cartItemList) {
            if (cartItem.getProduct().getId().equals(product.getId())) {
                if (cartItem.getQuantity() == 5) {
                    return false;
                }

                int index = cartItemList.indexOf(cartItem);
                cartItem.setQuantity(cartItem.getQuantity() + 1);
                cartItemList.set(index, cartItem);

                mutableCart.setValue(cartItemList);
                calculateCartTotal();
                return true;
            }
        }
        Cart cartItem = new Cart(product, 1);
        cartItemList.add(cartItem);
        mutableCart.setValue(cartItemList);
        calculateCartTotal();
        return true;
    }

    public void removeItemFromCart(Cart cartItem) {
        if (mutableCart.getValue() == null) {
            return;
        }
        List<Cart> cartItemList = new ArrayList<>(mutableCart.getValue());
        cartItemList.remove(cartItem);
        mutableCart.setValue(cartItemList);
        calculateCartTotal();
    }

    public  void changeQuantity(Cart cartItem, int quantity) {
        if (mutableCart.getValue() == null) return;

        List<Cart> cartItemList = new ArrayList<>(mutableCart.getValue());

        Cart updatedItem = new Cart(cartItem.getProduct(), quantity);
        cartItemList.set(cartItemList.indexOf(cartItem), updatedItem);

        mutableCart.setValue(cartItemList);
        calculateCartTotal();
    }

    private void calculateCartTotal() {
        if (mutableCart.getValue() == null) return;
        double total = 0.0;
        List<Cart> cartItemList = mutableCart.getValue();
        for (Cart cartItem: cartItemList) {
            total += cartItem.getProduct().getPrice() * cartItem.getQuantity();
        }
        mutableTotalPrice.setValue(total);
    }

    public LiveData<Double> getTotalPrice() {
        if (mutableTotalPrice.getValue() == null) {
            mutableTotalPrice.setValue(0.0);
        }
        return mutableTotalPrice;
    }

}

File: app/src/main/java/com/mesada/shop/repository/CartRepository.kt
package com.mesada.shop.repository

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.mesada.shop.model.Cart
import com.mesada.shop.model.Product
import java.util.*

class CartRepository {
    private val mutableCart = MutableLiveData<List<Cart>?>()
    private val mutableTotalPrice = MutableLiveData<Double?>()
    val cart: LiveData<List<Cart>?>
        get() {
            if (mutableCart.value == null) {
                initCart()
            }
            return mutableCart
        }

    fun initCart() {
        mutableCart.value = ArrayList()
        calculateCartTotal()
    }

    fun addItemToCart(product: Product?): Boolean {
        if (mutableCart.value == null) {
            initCart()
        }
        val cartItemList: MutableList<Cart> = ArrayList(mutableCart.value)
        for (cartItem in cartItemList) {
            if (product != null) {
                if (cartItem.product!!.id == product.id) {
                    if (cartItem.quantity == 5) {
                        return false
                    }
                    val index = cartItemList.indexOf(cartItem)
                    cartItem.quantity = cartItem.quantity + 1
                    cartItemList[index] = cartItem
                    mutableCart.value = cartItemList
                    calculateCartTotal()
                    return true
                }
            }
        }
        val cartItem = Cart(product, 1)
        cartItemList.add(cartItem)
        mutableCart.value = cartItemList
        calculateCartTotal()
        return true
    }

    fun removeItemFromCart(cartItem: Cart?) {
        if (mutableCart.value == null) {
            return
        }
        val cartItemList: MutableList<Cart> = ArrayList(mutableCart.value)
        cartItemList.remove(cartItem)
        mutableCart.setValue(cartItemList)
        calculateCartTotal()
    }

    fun changeQuantity(cartItem: Cart?, quantity: Int) {
        if (mutableCart.value == null) return
        val cartItemList: MutableList<Cart> = ArrayList(mutableCart.value)
        val updatedItem = Cart(cartItem!!.product, quantity)
        cartItemList[cartItemList.indexOf(cartItem)] = updatedItem
        mutableCart.value = cartItemList
        calculateCartTotal()
    }

    private fun calculateCartTotal() {
        if (mutableCart.value == null) return
        var total = 0.0
        val cartItemList = mutableCart.value
        for (cartItem in cartItemList!!) {
            total += cartItem.product!!.price * cartItem.quantity
        }
        mutableTotalPrice.value = total
    }

    val totalPrice: LiveData<Double?>
        get() {
            if (mutableTotalPrice.value == null) {
                mutableTotalPrice.value = 0.0
            }
            return mutableTotalPrice
        }
}

